_posts/2021-04-05-maven-bom.md

```diff
@@ -0,0 +1,312 @@
+
+## Managing Transitive Dependencies with Maven
+
+Now that we have an overview of the different dependency types in Maven, let's see in detail how Maven deals with transitive dependencies in a large-scale project.
+
+As an example, we will deal with Spring Framework and we will take two dependencies: `spring-context` and `spring-security-web`.
+
+In the POM file we add the direct dependencies:
+
+```xml
+<dependencies>
+    <dependency>
+        <groupId>org.springframework</groupId>
+        <artifactId>spring-context</artifactId>
+        <version>5.3.5</version>
+    </dependency>
+    <dependency>
+        <groupId>org.springframework.security</groupId>
+        <artifactId>spring-security-web</artifactId>
+        <version>5.4.5</version>
+    </dependency>
+</dependencies>
+```
+
+For someone who is not aware of transitive dependencies, he will think that using this dependency declaration, only two jars will be pulled. Fortunately, Maven provides a command that will show us what was pulled exactly concerning these two dependencies.
+
+Using the command `mvn dependency:tree`, we can list all the dependencies including the transitive ones. The result is:
+
+```markdown
++- org.springframework:spring-context:jar:5.3.5:compile
+|  +- org.springframework:spring-aop:jar:5.3.5:compile
+|  +- org.springframework:spring-beans:jar:5.3.5:compile
+|  +- org.springframework:spring-core:jar:5.3.5:compile
+|  |  \- org.springframework:spring-jcl:jar:5.3.5:compile
+|  \- org.springframework:spring-expression:jar:5.3.5:compile
+\- org.springframework.security:spring-security-web:jar:5.4.5:compile
+   +- org.springframework.security:spring-security-core:jar:5.4.5:compile
+   \- org.springframework:spring-web:jar:5.2.13.RELEASE:compile
+```
+
+It is a surprise: We started from two dependencies, and in this output, we find out that maven pulled seven other dependencies with different versions. The other dependencies are simply transitive. Just imagine the dependency tree that we will have in a multi-module project with hundreds of dependencies.
+
+The question that we can have now is how does Maven handle the case where multiple versions of the same dependency are pulled? 
+
+The first thing to know is that Maven can't sort versions: The versions are arbitrary strings and may not follow a strict semantic sequence. For example, if we have two versions `1.2` and `1.11`, we know that `1.11` comes after `1.2` but the String comparison gives `1.11` before `1.2`. Other version values can be "`1.1-rc1`" or "`1.1-FINAL`", that's why sorting versions by Maven is not a solution.  
```

Comment:
Before going into the details of dependency resolution you need to explain the problem with the above dependency tree.  Is there a problem with transitive dependencies in the dependency tree? I don't see one. Maybe start directly with the "-verbose" example to showcase the conflicts?